File: backend/src/application/events/events.service.ts
import { Injectable } from '@nestjs/common'
import type { Prisma } from '@prisma/client'
import { PrismaService } from '../../infra/prisma/prisma.service'
import type { CreateEventDto, EventCountDto } from '../../contracts/dto/event.dto'

@Injectable()
export class EventsService {
  constructor(private readonly prisma: PrismaService) {}

  async track(input: CreateEventDto): Promise<{ ok: true }> {
    await this.prisma.event.create({
      data: {
        name: input.name,
        props: (input.props ?? undefined) as Prisma.InputJsonValue | undefined,
        placeId: input.placeId,
      },
    })
    return { ok: true }
  }

  // Grouped counts for the admin dashboard.
  async counts(): Promise<EventCountDto[]> {
    const rows = await this.prisma.event.groupBy({
      by: ['name'],
      _count: { _all: true },
      orderBy: { _count: { name: 'desc' } },
    })
    return rows.map((r) => ({ name: r.name, count: r._count._all }))
  }
}

File: api/src/main/java/ee/gemspot/api/service/EventsService.java
package ee.gemspot.api.service;

import ee.gemspot.api.domain.Event;
import ee.gemspot.api.dto.CreateEventDto;
import ee.gemspot.api.dto.EventCountDto;
import ee.gemspot.api.repository.EventRepository;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Byte-identical port of {@code events.service.ts}.
 */
@Service
public class EventsService {

    private final EventRepository eventRepo;

    public EventsService(EventRepository eventRepo) {
        this.eventRepo = eventRepo;
    }

    public void track(CreateEventDto input) {
        Event event = new Event();
        event.setName(input.name());
        event.setProps(input.props());
        event.setPlaceId(input.placeId());
        eventRepo.save(event);
    }

    /** Grouped counts for the admin dashboard, ordered by count desc. */
    public List<EventCountDto> counts() {
        Map<String, Long> grouped = new LinkedHashMap<>();
        for (Event e : eventRepo.findAll()) {
            grouped.merge(e.getName(), 1L, Long::sum);
        }
        List<EventCountDto> out = new ArrayList<>();
        for (Map.Entry<String, Long> entry : grouped.entrySet()) {
            out.add(new EventCountDto(entry.getKey(), entry.getValue()));
        }
        out.sort(Comparator.comparingLong(EventCountDto::count).reversed());
        return out;
    }
}
